src/awkward/contents/recordarray.py

```diff
@@ -179,53 +186,50 @@ def __init__(
             backend = NumpyBackend.instance()
 
         if length is None:
+            # Require a length if we have no contents
             if len(contents) == 0:
                 raise TypeError(
                     "{} if len(contents) == 0, a 'length' must be specified".format(
                         type(self).__name__
                     )
                 )
 
-            if backend.nplike.known_data:
-                for content in contents:
-                    assert content.length is not unknown_length
-                    # First time we're setting length, and content.length is not unknown_length
-                    if length is None:
-                        length = content.length
-                    # length is not unknown_length, content.length is not unknown_length
-                    else:
-                        length = min(length, content.length)
-            else:
-                for content in contents:
-                    # First time we're setting length, and content.length is not unknown_length
-                    if length is None:
-                        length = content.length
-                        # Any unknown_length means all unknown_length
-                        if length is unknown_length:
-                            break
-                    # `length` is set, can't be unknown_length
-                    elif content.length is unknown_length:
-                        length = unknown_length
+            # Take length as minimum length of contents. This will touch shapes
+            it_contents = iter(contents)
+            for content in it_contents:
+                # First time we're setting length, and content.length is not unknown_length
+                if length is None:
+                    length = content.length
+                    # Any unknown_length means all unknown_length
+                    if length is unknown_length:
                         break
-                    # `length` is set, can't be unknown_length
-                    else:
-                        length = min(length, content.length)
+                # `length` is set, can't be unknown_length
+                elif content.length is unknown_length:
+                    length = unknown_length
+                    break
+                # `length` is set, can't be unknown_length
+                else:
+                    length = min(length, content.length)
+
+            # Touch everything else
+            for content in it_contents:
+                content._touch_shape(False)
```

Comment:
Before hand, we'd early-exit, and *not* touch all the shapes. But at runtime, we'll want to know all the shapes so we can pick the smallest.

Reply:
Right. (And early exits are dangerous for typetracer code.)